common/utils/demangle.hpp

```diff
@@ -0,0 +1,49 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOSTOOLS_COMMON_UTILS_DEMANGLE_HPP
+#define KOKKOSTOOLS_COMMON_UTILS_DEMANGLE_HPP
+
+#include <memory>
+#include <string>
+
+#if defined(__GXX_ABI_VERSION)
+#define HAVE_GCC_ABI_DEMANGLE
+#endif
+
+#if defined(HAVE_GCC_ABI_DEMANGLE)
+#include <cxxabi.h>
+#endif  // HAVE_GCC_ABI_DEMANGLE
+
+namespace KokkosTools {
+
+//! Demangle @p mangled_name.
+inline std::string demangleName(const std::string_view mangled_name) {
+#if defined(HAVE_GCC_ABI_DEMANGLE)
+  int status = 0;
+
+  std::unique_ptr<char, decltype(&std::free)> demangled_name(
+      abi::__cxa_demangle(mangled_name.data(), nullptr, nullptr, &status),
+      &std::free);
+
+  if (status == 0) return std::string(demangled_name.get());
+#endif
```

Comment:
Why bother with a `unique_ptr`?
```diff
diff --git a/common/utils/demangle.hpp b/common/utils/demangle.hpp
index 47cdfeb..b0ce5d9 100644
--- a/common/utils/demangle.hpp
+++ b/common/utils/demangle.hpp
@@ -17,7 +17,6 @@
 #ifndef KOKKOSTOOLS_COMMON_UTILS_DEMANGLE_HPP
 #define KOKKOSTOOLS_COMMON_UTILS_DEMANGLE_HPP

-#include <memory>
 #include <string>

 #if defined(__GXX_ABI_VERSION)
@@ -33,13 +32,14 @@ namespace KokkosTools {
 //! Demangle @p mangled_name.
 inline std::string demangleName(const std::string_view mangled_name) {
 #if defined(HAVE_GCC_ABI_DEMANGLE)
-  int status = 0;
-
-  std::unique_ptr<char, decltype(&std::free)> demangled_name(
-      abi::__cxa_demangle(mangled_name.data(), nullptr, nullptr, &status),
-      &std::free);
-
-  if (status == 0) return std::string(demangled_name.get());
+  int status;
+  char* demangled_name =
+      abi::__cxa_demangle(mangled_name.data(), nullptr, nullptr, &status);
+  if (demangled_name) {
+    std::string ret(demangled_name);
+    std::free(demangled_name);
+    return ret;
+  }
 #endif
   return std::string(mangled_name);
 }

Reply:
It wasn't crystal-clear to me from the documentation https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html if the caller would need to free the memory allocated by `abi::__cxa_demangle` if the status is not equal to zero:
> Returns:
    A pointer to the start of the NUL-terminated demangled name, or NULL if the demangling fails. The caller is responsible for deallocating this memory using free.

I just checked with *valgrind* that for a name that cannot be demangled, there is no leak.

Reply:
So I applied your suggestion :wink: